QA: stop nodes after RPC tests, even with --nocleanup

dexX7 opened the pull request:
`--nocleanup` should provide a way to preserve test data, but should not have an impact on whether nodes are to be stopped after the test execution.

In particular, when currently running RPC tests with `--nocleanup`, then it may result in several active `bitcoind`, which are not stopped properly.

~~Shutdown errors are ignored, given that a failure likely relates to already terminated processes, caused during the actual testing.~~

Edit:
Further, with `--noshutdown`, the nodes are not stopped explicitly. `--noshutdown` implies `--nocleanup`, to prevent removing datadirs, which are still in use.

dexX7 committed:
QA: stop nodes after RPC tests, even with --nocleanup

`--nocleanup` should provide a way to preserve test data, but should not have an impact on whether nodes are to be stopped after the test execution.

In particular, when currently running RPC tests with `--nocleanup`, then it may result in several active `bitcoind` processes, which are not terminated properly.

laanwj commented:
I was also surprised by the bitcoinds still running when specifying `--nocleanup`, it is not very intuitive. Though I'm sure that there can be reasons to want the nodes to remain running, for example to investigate their state (which is more difficult postmortem)

dexX7 commented:
> Though I'm sure that there can be reasons to want the nodes to remain running ... which is more difficult postmortem

There could be a `--noshutdown` option to explicitly prevent shutting down the nodes (which should imply `--nocleanup` to avoid removing datadirs that are still in use), but I think there would probably be not much use for it, due to the seemingly random selection of RPC ports.

Assuming there were such an option, additional reporting of the ports, or alternatively providing options to specify ports, or maybe even an option to run the QT client instead of bitcoind, would be helpful.

I'm not sure how other people use, or want to use, the tests, so it would be interesting to learn more about potential use cases. :)

laanwj commented:
You don't have to explicitly know the ports, you can use `bitcoin-cli -regtest -datadir=/tmp/<tempdir>/node0` to communicate with the instances.

laanwj commented on qa/rpc-tests/test_framework.py:
```diff
@@ -128,10 +128,14 @@ def main(self):
             stop_nodes(self.nodes)
             wait_bitcoinds()
+        except:
+            pass
```
Please report the exception w/ stack trace here (`traceback.print_exc()`). Ignoring exceptions, especially in tests, can cause useful diagnostic information to et lost.

dexX7 replied:
I was thinking: if the client crashed earlier, then stopping the nodes is going to fail, too, and this was the part I wanted to hide. I'm going to take a look at it. :)

dexX7 replied:
I might be wrong, but it looks like there doesn't seem to be an exception raised to begin with. I removed the catch completely in the updated version.

dexX7 committed:
QA: add --noshutdown option to prevent stopping nodes

With `--noshutdown`, the nodes are not stopped explicitly. `--noshutdown` implies `--nocleanup`, to prevent removing datadirs, which are still in use.

laanwj commented:
utACK

gavinandresen commented:
ACK

sipa commented:
Concept ACK. I didn't even know --nocleanup existed.

dgenr8 commented:
After-the-fact ACK.  --nocleanup is indispensable when something actually breaks, which is always when developing a test, but the zombie bitcoind's were a nuisance.
